[SPARK-25394][CORE] Add an application status metrics source

skonto opened the pull request:
## What changes were proposed in this pull request?

- Exposes several metrics regarding application status as a source, useful to scrape them via jmx instead of mining the metrics rest api.  Example use case: prometheus + jmx exporter.
- Metrics are gathered when a job ends at the AppStatusListener side, could be more fine-grained but most metrics like tasks completed are also counted by executors. More metrics could be exposed in the future to avoid scraping executors in some scenarios.
- a config option `spark.app.status.metrics.enabled` is added to disable/enable these metrics, by default they are disabled.

## How was this patch tested?

This was manually tested with jmx source enabled and prometheus server on k8s:
![metrics](https://user-images.githubusercontent.com/7945591/45300945-63064d00-b518-11e8-812a-d9b4155ba0c0.png)
In the next pic the job delay is shown for repeated pi calculation (Spark action).
![pi](https://user-images.githubusercontent.com/7945591/45329927-89a1a380-b56b-11e8-9cc1-5e76cb83969f.png)

Commit:
add application status source

skonto commented:
@felixcheung psl review.

skonto commented:
please test this

SparkQA commented:
**[Test build #95880 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95880/testReport)** for PR 22381 at commit [`b43190d`](https://github.com/apache/spark/commit/b43190d5393217f751f38149839a060df88baf1f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/2977/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/2978/
Test PASSed.

xuanyuanking commented on core/src/main/scala/org/apache/spark/SparkContext.scala:
```diff
@@ -560,6 +561,7 @@ class SparkContext(config: SparkConf) extends Logging {
 
     setupAndStartListenerBus()
     postEnvironmentUpdate()
+    _env.metricsSystem.registerSource(appStatusSource)
```
Better to put this line to +574

skonto replied:
Yeah I was wondering where that would be bet to put. So at the end makes sense on the other hand what if I want to have metrics as early as the app starts. Anyway, for now its ok.

xuanyuanking commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -280,6 +284,12 @@ private[spark] class AppStatusListener(
+      if (blacklisted) {
+        appStatusSource.foreach{_.BLACKLISTED_EXECUTORS.inc(1)}
+      }
+      else {
```
nit: } else {

skonto replied:
ok

xuanyuanking commented on core/src/main/scala/org/apache/spark/status/AppStatusSource.scala:
```diff
@@ -0,0 +1,55 @@
+
+import org.apache.spark.metrics.source.Source
+
+private[spark] class AppStatusSource extends Source{
```
nit: Source {

skonto replied:
:+1:

xuanyuanking commented on core/src/main/scala/org/apache/spark/status/AppStatusStore.scala:
```diff
@@ -503,9 +503,12 @@ private[spark] object AppStatusStore {
-  def createLiveStore(conf: SparkConf): AppStatusStore = {
+  def createLiveStore(
+      conf: SparkConf,
+      appStatusSource: Option[AppStatusSource] = None):
```
nit: `: AppStatusStore`?

skonto replied:
appStatusSource needs to be passed here as it is created in the SparkContext, do you mean something else?

squito replied:
I think the comment may be that that you can move `AppStatusStore = {` on to to same line

skonto replied:
Ok will do thanx.

xuanyuanking replied:
Yep, sorry for the late reply. :(

srowen commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -280,6 +284,12 @@ private[spark] class AppStatusListener(
     liveExecutors.get(execId).foreach { exec =>
       exec.isBlacklisted = blacklisted
+      if (blacklisted) {
+        appStatusSource.foreach{_.BLACKLISTED_EXECUTORS.inc(1)}
```
Nit: you don't need a block here, just parens. Same for several calls below.

skonto replied:
yeah I thought intellij complained, will fix.

srowen commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -382,11 +392,37 @@ private[spark] class AppStatusListener(
+      for {
+        submissionTime <- job.submissionTime
+        completionTime <- job.completionTime
+      } yield {
```
Nit: don't you technically need `foreach` here? I think it might be clearer to just wrap this in `if (job.submissionTime.isDefined && ...)` rather than the `for yield` that doesn't really loop.
Below -- isn't this just trying to take the difference in time in milliseconds between two dates? if so don't you just subtract their timestamps or is it more subtle here?

skonto replied:
Yeah I wanted to do this properly (java 8 wise, didnt think much of it) so on one hand I do have the timestamp on the other hand I do have the date object. So it was straightforward to pick the date. I didnt extract a timestamp out of the date object.

srowen replied:
`java.util.Date.getTime` gives you ms since the epoch directly, so I think you can subtract them and be done?

skonto replied:
It should be possible just tried the elegant java way: https://stackoverflow.com/questions/4927856/how-to-calculate-time-difference-in-java :)

skonto replied:
@srowen regarding multiple options its a valid technique: https://alvinalexander.com/scala/how-to-use-multiple-options-for-loop-comprehension
Anyway I can keep it simple and  just use ifDefined and then get the values, it liked the former though, more idiomatic.

srowen reviewed (commented):
I don't feel I know enough to really review. The thrust of this is _also_ logging metrics to JMX that are already shown in the metrics REST API? that mirroring of metrics seems like a bigger question or project.

skonto commented:
> that mirroring of metrics seems like a bigger question 
@srowen thanks for the comments, rest api is not always available or a standard way to export stuff.
This is also the experience coming from [Opsclarity](https://www.opsclarity.com/product/features/) work.
It it much more standard to use jmx for many systems out there AFAIK. This way using one component only like jmx exporter I can have all metrics in one place and no need to scrape the rest api.

skonto commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -382,11 +392,37 @@ private[spark] class AppStatusListener(
+          LocalDateTime.ofInstant(submissionTime.toInstant, ZoneId.systemDefault)
+        val localCompletionTime =
+          LocalDateTime.ofInstant(completionTime.toInstant, ZoneId.systemDefault)
+        val duration = Duration.between(localCompletionTime, localSubmissionTime)
```
need to reverse this.

skonto commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -382,11 +392,37 @@ private[spark] class AppStatusListener(
+        val localCompletionTime =
+          LocalDateTime.ofInstant(completionTime.toInstant, ZoneId.systemDefault)
+        val duration = Duration.between(localCompletionTime, localSubmissionTime)
+        appStatusSource.foreach{_.JOB_DURATION.update(duration.toMillis)}
```
Ideally this should be a gauge due to this: https://prometheus.io/docs/instrumenting/writing_exporters/#drop-less-useful-statistics

skonto replied:
Actually I will transform it to gauge makes more sense.

skonto commented on core/src/main/scala/org/apache/spark/status/AppStatusSource.scala:
```diff
@@ -0,0 +1,55 @@
+
+  val COMPLETED_STAGES = metricRegistry.counter(MetricRegistry.name("completedStages"))
+
+  val COMPLETED_JOBS = metricRegistry.counter(MetricRegistry.name("completedJobs"))
```
this is not used, need to remove.

SparkQA commented:
**[Test build #95880 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95880/testReport)** for PR 22381 at commit [`b43190d`](https://github.com/apache/spark/commit/b43190d5393217f751f38149839a060df88baf1f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #95881 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95881/testReport)** for PR 22381 at commit [`b43190d`](https://github.com/apache/spark/commit/b43190d5393217f751f38149839a060df88baf1f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95880/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95881/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/2991/
Test PASSed.

SparkQA commented:
**[Test build #95905 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95905/testReport)** for PR 22381 at commit [`a8fc89d`](https://github.com/apache/spark/commit/a8fc89d51971e16c37783cf336daa07d18e6d3c1).

SparkQA commented:
**[Test build #95905 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95905/testReport)** for PR 22381 at commit [`a8fc89d`](https://github.com/apache/spark/commit/a8fc89d51971e16c37783cf336daa07d18e6d3c1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95905/
Test PASSed.

skonto commented:
@srowen @xuanyuanking ready for another round.

srowen commented:
I don't think I know enough about monitoring to merge this. I am concerned about the duplication here, although I understand your argument why it's valuable.

skonto commented:
@srowen thanks is there someone I could call? @vanzin ?

skonto commented:
@srowen I was checking git history and I see that you merged this one: https://github.com/apache/spark/pull/22218 again related to metrics. Could you call someone who can do the merge?

srowen commented:
Yeah, this is bigger than adding one metric; raise it on dev@? I'm not sure who if anyone has an opinion on it. I'd rather hear more feedback on the history of these choices and what the right future state is. What's the use case, etc

erikerlandson commented:
These new metrics seem useful.  Is there a way to provide unit or integration testing for it?
Do these have enable/disable via metrics.properties files?

skonto commented:
@erikerlandson I dont any tests for sources AFAIK, they exist for the metric system itself. From the commit history this look like an old feature. Regarding enabling or disabling them no at the moment. But it is fairly easy to do via spark conf as there is one flag for structured streaming already `spark.sql.streaming.metricsEnabled`.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/3115/
Test PASSed.

SparkQA commented:
**[Test build #96076 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96076/testReport)** for PR 22381 at commit [`635f5aa`](https://github.com/apache/spark/commit/635f5aae4c91890c62b18edcf2b71400861cc463).

SparkQA commented:
**[Test build #96076 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96076/testReport)** for PR 22381 at commit [`635f5aa`](https://github.com/apache/spark/commit/635f5aae4c91890c62b18edcf2b71400861cc463).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96076/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/3167/
Test PASSed.

SparkQA commented:
**[Test build #96151 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96151/testReport)** for PR 22381 at commit [`677a7c3`](https://github.com/apache/spark/commit/677a7c3da10ae7c241a1b8fc5cb9a9768109c6b1).

SparkQA commented:
**[Test build #96151 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96151/testReport)** for PR 22381 at commit [`677a7c3`](https://github.com/apache/spark/commit/677a7c3da10ae7c241a1b8fc5cb9a9768109c6b1).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96151/
Test FAILed.

skonto commented:
@erikerlandson I added an option to disable them, by default they are now disabled. Is it possible to move forward with this? @srowen ? Btw I removed any code triggered by this unless explicitly enabled.

SparkQA commented:
**[Test build #96153 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96153/testReport)** for PR 22381 at commit [`3a2db16`](https://github.com/apache/spark/commit/3a2db16813a2bab3160f444fe0855855187a6178).

SparkQA commented:
**[Test build #96153 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96153/testReport)** for PR 22381 at commit [`3a2db16`](https://github.com/apache/spark/commit/3a2db16813a2bab3160f444fe0855855187a6178).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96153/
Test FAILed.

skonto commented:
test failure is irrelevant. jenkins, test this please.

skonto commented:
jenkins, test this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/3173/
Test PASSed.

SparkQA commented:
**[Test build #96159 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96159/testReport)** for PR 22381 at commit [`3a2db16`](https://github.com/apache/spark/commit/3a2db16813a2bab3160f444fe0855855187a6178).

SparkQA commented:
**[Test build #96159 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96159/testReport)** for PR 22381 at commit [`3a2db16`](https://github.com/apache/spark/commit/3a2db16813a2bab3160f444fe0855855187a6178).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96159/
Test FAILed.

skonto commented:
@srowen integration tests seem so flaky! Seen this before?

srowen commented:
Looks like different failures. I am not sure if these are known issues. I'd try again later today.

xuanyuanking commented:
UT fixed by #22452.

xuanyuanking commented:
retest this please.

SparkQA commented:
**[Test build #96215 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96215/testReport)** for PR 22381 at commit [`3a2db16`](https://github.com/apache/spark/commit/3a2db16813a2bab3160f444fe0855855187a6178).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/3211/
Test PASSed.

vanzin commented:
I don't really have an opinion about the metrics or names or things like that; but not being familiar with the internals of this metric system, I'd just like to make sure that updating these metrics is cheap and won't do things like actually try to send data to the sinks, which would be a no-no in this particular listener.

Also, if the metrics system can be disabled (isn't it disabled by default?), I'd avoid even updating these metrics in the listener at all.

skonto commented:
@vanzin AFAIK the operations just update metrics in the underlying dropwizard metrics library. I dont htink anything is shipped anywhere. How should I proceed? Is there anyone who is familiar with the internals?

aditanase commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -280,6 +284,11 @@ private[spark] class AppStatusListener(
     liveExecutors.get(execId).foreach { exec =>
       exec.isBlacklisted = blacklisted
+      if (blacklisted) {
+        appStatusSource.foreach(_.BLACKLISTED_EXECUTORS.inc(1))
```
You could just say `.inc()` - 1 is the default. Same everywhere else.

skonto replied:
ok

aditanase commented on core/src/main/scala/org/apache/spark/status/AppStatusSource.scala:
```diff
@@ -0,0 +1,81 @@
+  val JOB_DURATION = metricRegistry
+    .register(MetricRegistry.name("jobDuration"), jobDuration)
+
+  val FAILED_STAGES = metricRegistry.counter(MetricRegistry.name("failedStages"))
```
I would suggest creating groups of metrics by adding a prefix with the "context", e.g. `stages`, `jobs`, `tasks`.
This is a best practice followed by other sources. See:
- https://github.com/apache/spark/blob/9fed6abfdcb7afcf92be56e5ccbed6599fe66bc4/core/src/main/scala/org/apache/spark/scheduler/DAGSchedulerSource.scala#L29
- https://github.com/apache/spark/blob/a4491626ed8169f0162a0dfb78736c9b9e7fb434/core/src/main/scala/org/apache/spark/storage/BlockManagerSource.scala#L38

skonto replied:
ok makes sense

aditanase commented on core/src/main/scala/org/apache/spark/status/AppStatusSource.scala:
```diff
@@ -0,0 +1,81 @@
+
+  val SKIPPED_TASKS = metricRegistry.counter(MetricRegistry.name("skippedTasks"))
+
+  val BLACKLISTED_EXECUTORS = metricRegistry.counter(MetricRegistry.name("blackListedExecutors"))
```
Given that most metrics are counters, I would add a small inline function that reduces some of the repetition here, e.g.
```scala
def getCounter(prefix: String, name: String): Counter =
  metricRegistry.counter(MetricRegistry.name(prefix, name))
```

skonto replied:
+1

aditanase reviewed (commented):
Left a few suggestions, mostly cosmetic.

aditanase commented:
@skonto @vanzin You can rest assured the metrics are not shipped anywhere. Collection is cheap/free and completely decoupled from reporting.
The registry is only a data bag, shipping metrics is done via reporters that are configured to "pull" metrics from the registry at regular intervals.
Also, you're mostly adding counters that are extremely simple. Histograms and timers are more involved but the cost associated is far outweighed by the benefits of having data points.

skonto commented:
Thanx a lot @aditanase ! I will update the PR.

SparkQA commented:
**[Test build #97324 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97324/testReport)** for PR 22381 at commit [`86bf7ec`](https://github.com/apache/spark/commit/86bf7ec4bbe1bca268f85464b73aa0cdcb0bf163).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/3931/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97324/
Test PASSed.

skonto commented:
@vanzin could I get a merge pls?

skonto commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -382,11 +391,38 @@ private[spark] class AppStatusListener(
+      }
+
+      // update global app status counters
+      appStatusSource.foreach(_.COMPLETED_STAGES.inc(job.completedStages.size))
```
ok will fix.

skonto commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -382,11 +391,38 @@ private[spark] class AppStatusListener(
+            LocalDateTime.ofInstant(submissionTime.toInstant, ZoneId.systemDefault)
+          val localCompletionTime =
+            LocalDateTime.ofInstant(completionTime.toInstant, ZoneId.systemDefault)
+          val duration = Duration.between(localSubmissionTime, localCompletionTime)
```
Yeah just did the Java 8 thing, guess too much, will fix.

skonto commented on core/src/main/scala/org/apache/spark/status/AppStatusSource.scala:
```diff
@@ -0,0 +1,85 @@
+  def createSource(conf: SparkConf): Option[AppStatusSource] = {
+    Option(conf.get(AppStatusSource.APP_STATUS_METRICS_ENABLED))
+      .filter(identity)
+      .map(_ => new AppStatusSource())
```
ok :)

skonto commented:
@vanzin thanx for the review I think It looks ok now.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/4005/
Test PASSed.

SparkQA commented:
**[Test build #97415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97415/testReport)** for PR 22381 at commit [`a08836c`](https://github.com/apache/spark/commit/a08836cc9ca91164cf5dd95a63929d0613779cf2).

SparkQA commented:
**[Test build #97415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97415/testReport)** for PR 22381 at commit [`a08836c`](https://github.com/apache/spark/commit/a08836cc9ca91164cf5dd95a63929d0613779cf2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97415/
Test PASSed.

skonto commented:
@vanzin ready to merge.

skonto commented:
@vanzin didn't mean to press or anything :)

vanzin commented on core/src/main/scala/org/apache/spark/status/AppStatusListener.scala:
```diff
@@ -17,6 +17,7 @@
 
 package org.apache.spark.status
 
+import java.time.Duration
```
Unused?

vanzin commented on core/src/main/scala/org/apache/spark/status/AppStatusSource.scala:
```diff
@@ -0,0 +1,85 @@
+private[spark] object AppStatusSource {
+
+  def getCounter(prefix: String, name: String)(implicit metricRegistry: MetricRegistry): Counter = {
+    metricRegistry.counter (MetricRegistry.name (prefix, name) )
```
nit: no space before `(`

vanzin reviewed (commented):
Looks ok. I'll fix the two things during the merge.

Merging to master.

skonto commented:
Thanx @vanzin!
